test/src/test/java/jenkins/tasks/SimpleBuildWrapperTest.java

```diff
@@ -70,6 +84,61 @@
         }
     }
 
+    @Test public void envOverrideExpand() throws Exception {
+        Assume.assumeFalse(Functions.isWindows());
+        FreeStyleProject p = r.createFreeStyleProject();
+        p.getBuildWrappersList().add(new WrapperWithEnvOverrideExpand());
+        SpecialEnvSlave slave = new SpecialEnvSlave(tmp.getRoot(), r.createComputerLauncher(null));
+        r.jenkins.addNode(slave);
+        p.setAssignedNode(slave);
+        JDK jdk = new JDK("test", "/opt/jdk");
+        r.jenkins.getJDKs().add(jdk);
+        p.setJDK(jdk);
+        CaptureEnvironmentBuilder captureEnvironment = new CaptureEnvironmentBuilder();
+        p.getBuildersList().add(captureEnvironment);
+        p.getBuildersList().add(new Shell("echo effective PATH=$PATH"));
+        FreeStyleBuild b = r.buildAndAssertSuccess(p);
+        String expected = "/home/jenkins/extra/bin:/opt/jdk/bin:/usr/bin:/bin";
+        assertEquals(expected, captureEnvironment.getEnvVars().get("PATH"));
+        // TODO why is /opt/jdk/bin added twice? In CommandInterpreter.perform, envVars right before Launcher.launch is correct, but this somehow sneaks in.
+        r.assertLogContains("effective PATH=/opt/jdk/bin:" + expected, b);
+    }
+    public static class WrapperWithEnvOverrideExpand extends SimpleBuildWrapper {
+        @Override public void setUp(Context context, Run<?,?> build, FilePath workspace, Launcher launcher, TaskListener listener, EnvVars initialEnvironment) throws IOException, InterruptedException {
+            assertEquals("/opt/jdk/bin:/usr/bin:/bin", initialEnvironment.get("PATH"));
+            assertEquals("/home/jenkins", initialEnvironment.get("HOME"));
+            context.env("EXTRA", "${HOME}/extra");
+            context.env("PATH+EXTRA", "${EXTRA}/bin");
+        }
+        @TestExtension("envOverrideExpand") public static class DescriptorImpl extends BuildWrapperDescriptor {
+            @Override public String getDisplayName() {
+                return "WrapperWithEnvOverrideExpand";
+            }
+            @Override public boolean isApplicable(AbstractProject<?,?> item) {
+                return true;
+            }
+        }
+    }
+    private static class SpecialEnvSlave extends Slave {
```

Comment:
Because then I would get a plain `SlaveComputer`, whose `getEnvironment` will yield the environment variables of the computer on which this test is running. (The `EnvVars` parameter is passed to any `Launcher` the slave creates, but nothing else, including `initialEnvironment` above or the value from `CaptureEnvironmentBuilder`.) I wanted to make the test simulate a remote slave with different (and predictable) environment variables, to make sure that master variables are not accidentally “leaking” into the result.

A different approach would be to modify `JenkinsRule.createSlave` to pass the `EnvVars` not just to the `createComputerLauncher`, but to also define a `SpecialEnvSlave`/`SpecialEnvComputer` like I have here, for the benefit of other tests. There is some risk this would be a compatibility issue for existing test callers of `createSlave` (and similar methods) with a non-null environment, but perhaps it is worthwhile.